Feature/devops ci (#3799)
conan-io/docs: ci_tutorial/products_pipeline/distributed_build.rst

Products pipeline: distributed build
====================================

The previous section used ``--build=missing`` to build all the necessary packages in the same CI machine.

> This is not always desired, or even possible, and in many situations it is preferable to do a distributed build, to achieve faster builds and better use the CI resources. The most natural distribution of the build load is to build different packages in different machines. Let's see how this is possible with the ``conan graph build-order`` command.

Reviewer: ```suggestion
This is not always desired, or even possible, and in many situations it is preferable to do a distributed build, to achieve faster builds and better usage the CI resources. The most natural distribution of the build load is to build different packages in different machines. Let's see how this is possible with the ``conan graph build-order`` command.
```

Let's start as usual making sure we have a clean environment with the right repositories defined:

.. code-block:: bash

    # First clean the local "build" folder
    $ pwd  # should be <path>/examples2/ci/game
    $ rm -rf build  # clean the temporary build folder 
    $ mkdir build && cd build # To put temporary files

    # Now clean packages and define remotes
    $ conan remove "*" -c  # Make sure no packages from last run

    # If you did in previous sections, NO need to repeat this
    $ conan remote remove "*"  # Make sure no other remotes defined
    # Add products repo, you might need to adjust this URL
    # NOTE: The products repo is added first, it will have higher priority.
    $ conan remote add products http://localhost:8081/artifactory/api/conan/products
    # Add develop repo, you might need to adjust this URL
    $ conan remote add develop http://localhost:8081/artifactory/api/conan/develop

> We will obviate by now the ``mapviewer/1.0`` and focus this section in the ``game/1.0`` product.

Reviewer: ```suggestion
We will obviate by now the ``mapviewer/1.0`` product and focus this section in the ``game/1.0`` product.
```

> The first step is to compute the "build-order", that is, the list of packages that need to be build, and in what order.

Reviewer: ```suggestion
The first step is to compute the "build-order", that is, the list of packages that need to be built, and in what order.
```

This is done with the following ``conan graph build-order`` command:

.. code-block:: bash

>     $ conan graph build-order --requires=game/1.0 --build=missing 

Author: Done

      --order-by=recipe --reduce --format=json > game_build_order.json

Note a few important points:

- It is necessary to use the ``--build=missing``, in exactly the same way than in the previous section. Failing to provide the intended ``--build`` policy and argument will result in incomplete or erroneous build-orders.

> - The ``--reduce`` eliminates all elements in the result that doesn't have the ``binary: Build`` policy. This means that the resulting "build-order" cannot be merged with other build order files for aggregating them into a single one, which is important when there are multiple configurations and products.

Reviewer: ```suggestion
- The ``--reduce`` argument eliminates all elements in the resulting order that don't have the ``binary: Build`` policy. This means that the resulting "build-order" cannot be merged with other build order files for aggregating them into a single one, which is important when there are multiple configurations and products.
```

> - The ``--order-by`` argument allows to define different orders, by "recipe" or by "configuration". In this case, we are using the ``--order-by=recipe`` which is intended to parallelize builds per recipe, that means, that all possible different binaries for a given package like ``engine/1.0`` should be built first before any consumer of ``engine/1.0`` can be built.

Reviewer: ```suggestion
- The ``--order-by`` argument allows to define different orders, by "recipe" or by "configuration". In this case, we are using ``--order-by=recipe`` which is intended to parallelize builds per recipe, that means, that all possible different binaries for a given package like ``engine/1.0`` should be built first before any consumer of ``engine/1.0`` can be built.
```

                                "package_id": "bac7cd2fe1592075ddc715563984bbe000059d4c",
                                "binary": "Build",
                                "build_args": "--requires=game/1.0 --build=game/1.0",
                            }
                        ]
                    ]
                }
            ]
        ]
    }

For convenience, in the same way that ``conan graph info ... --format=html > graph.html`` can generate a file with an HTML interactive dependency graph, the ``conan graph build-order ... --format=html > build_order.html`` can generate an HTML visual representation of the above json file:

.. image:: ./build_order_simple.png
   :width: 500 px
   :align: center

The resulting json contains an ``order`` element which is a list of lists. This arrangement is important, every element in the top list is a set of packages that can be built in parallel because they do not have any relationship among them. You can view this list as a list of "levels", in level 0, there are packages that have no dependencies to any other package being built, in level 1 there are packages that contain dependencies only to elements in level 0 and so on.

> Then, the order of the elements in the top list is important and must be respected. Until the build of all the packages in one list item has finished, it is not possible to start the build of the next "level".

Reviewer: ```suggestion
Then, the order of the elements in the outermost list is important and must be respected. Until the build of all the packages in one list item has finished, it is not possible to start the build of the next "level".
```

> Using the information in the ``graph_build_order.json`` file, it is possible to execute the build of the necessary packages, in the same way that the previous section ``--build=missing`` did, but not directly managed by us.

Reviewer: ```suggestion
Using the information in the ``graph_build_order.json`` file, it is possible to execute the build of the necessary packages, in the same way that the previous section's ``--build=missing`` did, but not directly managed by us.
```

Taking the arguments from the json, the commands to execute would be:

.. code-block:: bash

    $ conan install --requires=engine/1.0 --build=engine/1.0
    $ conan install --requires=game/1.0 --build=game/1.0

> We are executing these commands manually, but in practice, it would be a ``for`` loop in CI executing over the json output. We will see some Python code later for this. At this point we wanted to focus on the ``conan graph build-order`` command, but we haven't really explained how the build is distributed.

Reviewer 2: Maybe some re-ordering of these parts is in order, since we name it `distributed build` but do not in fact explain it

Author: I have renamed this section, now it is called "the build-order", which I think is better than "distributed build", because it is a bit too early to introduce that concept here, and it is delayed until last scenario.
